stats/opentelemetry/internal/tracing/custom_carrier.go

```diff
@@ -0,0 +1,103 @@
+/*
+ *
+ * Copyright 2024 gRPC authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+
+// Package tracing implements the OpenTelemetry carrier for context propagation
+// in gRPC tracing.
+package tracing
+
+import (
+	"context"
+	"errors"
+
+	"go.opentelemetry.io/otel/propagation"
+	"google.golang.org/grpc/metadata"
+	"google.golang.org/grpc/stats"
+)
+
+// GRPCTraceBinHeaderKey is the gRPC metadata header key `grpc-trace-bin` used
+// to propagate trace context in binary format.
+const GRPCTraceBinHeaderKey = "grpc-trace-bin"
+
+// CustomCarrier is a TextMapCarrier that uses gRPC context to store and
+// retrieve any propagated key-value pairs in text format along with binary
+// format for `grpc-trace-bin` header
+type CustomCarrier struct {
+	propagation.TextMapCarrier
+
+	Ctx *context.Context
+}
+
+// NewCustomCarrier creates a new CustomMapCarrier with
+// the given context.
+func NewCustomCarrier(ctx context.Context) CustomCarrier {
+	return CustomCarrier{
+		Ctx: &ctx,
```

Comment:
The type context.Context is an interface type, which means it's zero value is nil and it is a pointer. Why are you storing a pointer to a pointer, and then deference the pointer to pointer? Why not just store the pointer as is, and then you don't have to deref. We usually don't store a pointer to a pointer unless you need to access the pointer through an atomic due to performance concerns/it's on a fast path/you want to atomically swap to a whole new configuration: https://github.com/grpc/grpc-go/blob/master/xds/internal/server/conn_wrapper.go#L70.

Reply:
I see. yeah, this is not idiomatic in Go. I am storing a pointer to context.Context (*context.Context) because my understanding is that we want to update the context in place as the initial context might have some existing keys that we should retain when returning new context from `Set` and `SetBinary` or do you think we are not required to that and rely on caller to merge them?

If we do require to do that, we can achieve this by introducing a mutable wrapper around the context. Instead of storing a pointer to context.Context, we'll store a pointer to a custom contextHolder struct that contains the context. This way, even though our methods have value receivers, they can modify the context inside the contextHolder, and these changes will be visible outside the methods. Btw,  we can't switch our implementation to pointer receiver i.e. `func (c *CustomCarrier) Set(key, value string)` Since `otelpropagation.TextMapCarrier` requires methods with value receivers.

Reply:
Let me know what you think

Reply:
I don't understand what you are referring to here. Persisting a pointer which points to a pointer (what an interface type is) which points you to some data has nothing technically distinct between the two options that to do with "update the context in place" or merging. Regardless of whether you store a pointer to a pointer or a pointer the same context operations are available to you, so storing a pointer to a pointer has no value add here, expect making it more complicated where everytime you access this field you dereference the pointer to a pointer to a pointer. Whether or not the custom carrier's receiver is a pointer or not has nothing to do with what fields it stores in its struct.

Reply:
The initial context passed to `NewCustomCarrier` might contain keys set by the caller. When Set or SetBinary are called, I need to update the context while retaining these existing keys. Modifying the context in place, through a pointer, allowed me to do that. 

Anyways, i have modified the implementation to using pointer receivers which is a simpler and more idiomatic solution for our `CustomCarrier`. It eliminates the need for a pointer to a pointer and aligns better with typical Go practices.

Reply:
Updating a field doesn't require it to be a pointer to a pointer. If you're writing a new value to an interface (which is already a pointer) you can do type.interfaceField = newThing. Modifying something doesn't require a pointer to a pointer, I'm not following what you're saying.

Reply:
`c.ctx = context.NewOutgoingContext(ctx, md)` was not retaining the existing keys in ctx metadata but `*c.ctx = context.NewOutgoingContext(ctx, md)` was. Anyways, now that I have changed the implementation to pointer receiver, the existing keys are being retained without using pointer to pointer and it looks more simpler